AdyenNetworking/APIClient/APIClient.swift

```diff
@@ -121,21 +130,30 @@ public final class APIClient: APIClientProtocol, AsyncAPIClientProtocol {
     private static func handle<R: Request>(
         _ result: URLSessionSuccess,
         _ request: R
-    ) throws -> R.ResponseType {
+    ) throws -> HTTPResponse<R.ResponseType> {
+        log(result: result, request: request)
         do {
-            log(result: result, request: request)
-            return try Coder.decode(result.data) as R.ResponseType
-        } catch {
-            if let errorResponse: R.ErrorResponseType = try? Coder.decode(result.data) {
-                throw errorResponse
-            } else if (200...299).contains(result.response.statusCode) == false {
-                throw HttpError(
-                    errorCode: result.response.statusCode,
-                    errorMessage: "Http \(result.response.statusCode) error"
+            if (200...299).contains(result.statusCode) {
+                return HTTPResponse(
+                    headers: result.headers,
+                    statusCode: result.statusCode,
+                    response: try Coder.decode(result.data) as R.ResponseType
                 )
             } else {
-                throw error
+                throw HTTPError(
+                    headers: result.headers,
+                    statusCode: result.statusCode,
+                    errorResponse: try Coder.decode(result.data) as R.ErrorResponseType
+                )
```

Comment:
the problem with this is that it can be status code 200 and the response body is `R.ErrorResponseType`, which is bad API design from the backend side, but this is the reality of it 🤷 .

that's why the check `if let errorResponse: R.ErrorResponseType = try? Coder.decode(result.data)` was done before checking the response code.

Reply:
makes sense. i'll revert this logic